### ISSUE TITLE
multiple languages on the braille display

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#2044]
**Reported by fatma.mehanna on 2012-01-17 11:34**
hi list,
nvda doesn't allow multiple languages to be displayed on braille display.if i set my braille display table to arabic 
for example,
and during reading an arabic text
i found suddenly a hebrew word for example,
 nvda doesn't display it correctly.

if i set my braille table to arabic for example,

**Comment by dkager** [comment]
I would say this needs to be addressed in liblouis. Switching braille tables isn't a good idea because it assumes that people learn to read a language's native braille code. I don't think this is true for most people. For example, I don't read UEB but definitely want to read English texts. Even if you want to read a language's alphanumeric characters in the native braille code, such as the ä and ö in Finnish, you probably don't know all the punctuation signs of that code.

So in short, I think the braille table should provide the dot patterns for these characters. For example, JAWS has dot patterns for Greek and many other Unicode ranges in its computer braille table.

**Comment by jcsteh** [comment]
I think there's some merit in supporting this but making it configurable. There's an analogy in speech: some people might choose to read English with a synthesiser that primarily speaks another language because the accent might be more understandable to them (assuming the synth handles this well enough). So, I don't think this should be a won't fix, but it's definitely a low priority feature.

**Comment by dkager** [comment]
One more thing to keep in mind is that a word in another language that occurs in the middle of a sentence will probably break capital sign rules from the braille table. Liblouis has to switch to a different table, which results in a different set of capital sign rules. So this idea is most interesting for people who are proficient in two or more codes.

**Comment by PratikP1** [comment]
One significant reason to add this type of functionality is for people needing to access textbooks when learning languages. These often commingl two languages. It requires that the book is properly marked up. I'll ask someone at the DAISY Consortium or Epub forum to get a sample document for a test case.

**Comment by Adriani90** [comment]
@andre9642 has written code for this. In my view, people who read multilingual texts know already the native braille of the languages they want to read in. So it should be an optional setting. I also have creadet an idea on Liblouis project for an universal 8 dot computer braille for every language which would require users to learn reading like that. But I think Andre's solution is also very good.

@andre9642 do you plan to raise a pull request for NVDA?

@leonardder your thoughts are appreciated here.

**Comment by Andre9642** [comment]
Hi @Adriani90,
My code for this is *very basic*. I have another code (currently not included in my add-on) which allows multiple outputs braille through auto-detection of the language (provided that the document is well tagged) but this code is horrible and not optimized...
In short, if it's for an integration in NVDA Core, I prefer to let other people more competent to do it.
Honestly, I still have a lot of work to make good code for something serious.

Regarding an universal 8 dot computer braille, I think that will be great but... a bit utopian. With 256 combinations, whether on one, two or three cells, we can never represent all signs...

**Comment by zstanecic** [comment]
Regarding an universal 8 dot computer braille, I think that will be 
great but... a bit utopian. With 256 combinations, whether on one, two 
or three cells, we can never represent all signs...
i agree with the above sentence from andre,
as liblouis has only basic imput/output mapping

**Comment by Adriani90** [comment]
cc: @dineshkaushal

**Comment by yeatersink** [comment]
@LeonarddeR  is there anyway that we can work on trying to enable various Braille tables or multiple rail tables in NVDA at the same time so that we can get multiple language support in the same document?

**Comment by LeonarddeR** [comment]
In my opinion, a user experience like this would be changing the list of braille tables into a checkable list. That allows you to enable/disable tables, but it will also require the ability to move tables up and down, because ordering of the tables matters.
This is also why I believe with @dkager that the primary responsibility for such functionality should be on liblouis' side.

One major problem with how many tables are currently designed, is that they usually include characters that are strictly spoken unrelated to the main language of the table.
For example, speaking about your (@yeatersink) hebrew table, that also includes the latin alphabet and some punctuation. Imagine what would happen if one enables both Hebrew and UEB. The order of these tables will matter, I assume the first table should be UEB. But it is hard to say what some of the more complex UEB rules will do with the output of Hebrew. So long story short, liblouis is currently not accustomed for this.

**Comment by LeonarddeR** [comment]
Looking at this issue as well as https://github.com/liblouis/liblouis/pull/1610#issuecomment-2497402776, it might be interesting to prototype the checkable list approach.
@AAClause How does you add-on work with multiple languages?

**Comment by Adriani90** [comment]
I've used the multiple languages feature of braille extender when i had a Braille display at hand, and it worked quite well with languages that had the same braille table (e.g. 8 dot computer braille for german and english). However, in my experience it was kind of a mess when two languages were mixed with different braille output tables.

But maybe it was my own inability to coordinate reading 6 dot braille in one language and 8 dot braille in another language on the same line. :)) I guess if people know well the braille patern of the languages the text is written in, they will very probable not have a problem in reading those languages on the braille display when they are shown at the same time.

I think though following UX might be nice:

1. In Braille settings, the braille input and output table comboboxes are redesigned to a tree view
2. In this tree view, Every_ language is an item which can be expanded or colapsed
3. When expanded, the braille tables for the language can be enabled either via context menu, or with space bar if they are checkboxes within the tree view.

It seems a combined tree view with check boxes and even with a context menu on each item can be created in WX Python 4.2.2:
https://docs.wxpython.org/wx.dataview.TreeListCtrl.html
It supports multi select as well.
